Rewrite factorio server spawning and communication

Hornwitser opened the pull request:
This is a do over of the server spawning and communication logic, the idea is to get some structure on this instead of being a complete mess in client.js.  It's partly necessary for #200 as it proved to be problematic to adopt the new configuration system with all the clutter present.  It's also a hard requirement for #201.

I've not yet completed this, but I wanted to put this out in case anyone has some thoughts they would like to share on this.

psihius commented on client.js:
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
 	}
 }
 
+/**
```
Question, maybe this should go into the lib folder as utility methods? So we don't have big fat client.js with all the code and something more manageable?

psihius commented on client.js:
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
+	}
+}
+
 async function createInstance(config, args, instance) {
```
Same goes here. Also, this function is somewhat big, ideally you split it into smaller chunks and that also lends itself to go into a file of its own or something :)

Hornwitser replied:
Unless you have some grand architectural structure pre planned, I feel like it's the wrong time to be moving these functions around.

psihius commented on client.js:
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
+	await fs.symlink(path.relative(path.dirname(link), target), link, 'junction');
+
+	await symlinkMods(instance, "sharedMods", console);
 	let instconf = {
```
Maybe use actual files with default values, so they are not defined in the code like this and loaded via require?

psihius commented on client.js:
```diff
@@ -362,11 +315,16 @@ write-data=${ instance.path() }\r\n
 	}
-	let serversettings = {
+
+	let overrides = {
```
Same here - load it from file?

psihius commented on client.js:
```diff
@@ -375,65 +333,55 @@ write-data=${ instance.path() }\r\n
+		serverSettings[name] = value;
+	}
+
+	await fs.writeFile(instance.path("server-settings.json"), JSON.stringify(serverSettings, null, 4));
```
filename should be a `const server_config = "server-settings.json";`  or something like that. We have been working on removing hard-coded file names from the codebase.

psihius replied:
The thing is - you define the file names somewhere at the top of the file, preferably the root file and you just use that variable throughout the code. And it makes it easier to also add the ability to pass the file name as a command-line parameter. While we use instance-per-dir approach, you might want to use more than one server-settings.json file or even store them somewhere else and pass them at launch as paths.

psihius commented on client.js:
```diff
@@ -796,7 +680,7 @@ async function instanceManagement(config, instance, instanceconfig) {
 		getID();
 		function getID() {
-			messageInterface("/silent-command rcon.print(#game.connected_players)", function(playerCount) {
+			server.sendRcon("/silent-command rcon.print(#game.connected_players)").then((playerCount) => {
```
But are we using only RCON going forward? stdin/stdout is a valid communication layer too and for some things probably is even better.

Hornwitser commented:
I've completed most of the code now.  Summary of changes:
- Refactored fileOps.getNewestFile as it was not compatible with the idea that you pass the name of the file as it appears in the saves directory.
- Renamed factorioPort to gamePort in the class.
- Removed some commented out code.
- Refactored getFactorioLocale and moved it to lib/factorio/locale
- Used the ini module to create the config.ini for Factorio
- Introduced a modular outputHeuristics filter which triggers actions based on outputs of Factorio.
- Changed the start and startScenario to wait until the server is ready to accept clients before returning.
- Added the disableAchievements method.
- Added integration tests.

Hornwitser commented:
I've completed this now.  Latest changes are that I've removed the deriving of instance id from the RCON password, removed stringUtils, and fixed some issues with shutdown and rcon on Windows.

Danielv123 commented:
Your changelog is missing the fact that instanceconfig.unique is now deprecated in favour of instanceconfig.id - also, other places in the code its called instanceID, isn't it? Maybe consider changing that if it makes sense

Hornwitser commented:
I'm pretty sure the config rewrite is going to have change this again so keeping it as unique should be fine for now.  I didn't want to name it unique in the config file and calling it instanceconfig.instanceID seems a little redundant.
